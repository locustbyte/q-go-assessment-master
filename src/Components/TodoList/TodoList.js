import React from 'react';
import uuid from 'uuid';
import moment from 'moment';
import Instructions from '../Instructions/index';
import TodoAdd from '../TodoAdd/TodoAdd';
import TodoItem from '../TodoItem/TodoItem';

class TodoList extends React.Component {

    filters = {
        ALL: 'ALL',
        COMPLETE: 'COMPLETE',
        INCOMPLETE: 'INCOMPLETE'
    }

    constructor() {
        super();
        this.state = {
            todos: this.initTodos(),
            filter: this.filters.ALL
        };

        this.toggleDone = this.toggleDone.bind(this);
        this.removeTodo = this.removeTodo.bind(this);
    }

    initTodos() {
        return [
            {
                title: "Make sure items are completeable",
                completed: false,
                added: moment("2017-09-29").toISOString(),
                
                id: uuid.v4()
            },
            {
                title: "Add filters (Use HOC)",
                completed: false,
                added: moment("2017-09-29").toISOString(),
                
                id: uuid.v4()
            }
        ];
    }

    

    removeTodo(name, i){
        let todos = this.state.todos.slice();
        
        for(var ii = 0; ii < todos.length; ii += 1) {
            if(todos[ii].id === name) {
                
                todos.splice(ii, 1);
                this.setState({
                    todos
                });
                return ii;
            }
        }
        return -1;        
    }
    

    addTodo(title) {
        const todo = {
            title,
            completed: false,
            added: moment().toISOString(),
            
            id: uuid.v4()
        };
        this.setState({todos: this.state.todos.concat(todo)})
    }

    toggleDone(id) {
        console.log(id)
        this.setState({
            todos: this.state.todos.map(t => {
                t.completed = t.id === id ? !t.completed : t.completed
                return t;
            })
        })
    }

    renderTodos() {
        return this.state.todos.filter(v => {
            switch (this.state.filter) {
                case this.filters.COMPLETE:
                    return v.completed === true;
                case this.filters.INCOMPLETE:
                    return v.completed === false;
                default:
                    return true;
            }
        }).map(v =>
            <TodoItem
                key={v.id}
                
                todo={v} toggleDone={this.toggleDone} removeTodo={this.removeTodo}
                
            />
        );
    }

    handleFilter(filter) {
        this.setState({
            filter: this.filters[filter]
        });
    }

    renderFilters() {
        return Object.keys(this.filters).map(v =>
            <li
                key={v}
                className={this.state.filter === v ? "active" : ""}
                onClick={this.handleFilter.bind(this, v)}>
                {v}
            </li>
        );
    }

    render() {
        return (
            <div className={'itemCreator'}>
                <TodoAdd
                    addTodo={this.addTodo.bind(this)}
                    
                />
                <ul className={'itemsList-ul'}>
                    {this.renderTodos()}
                </ul>
                
                <ul className="filters">
                    <li>Showing: </li>
                    {this.renderFilters()}
                </ul>
                <Instructions />

            </div>
        );
    };
}

TodoList.defaultProps = {
    todo: {
        title: "Empty todo",
        
        completed: false
    },
    
};

TodoList.propTypes = {
    todo: React.PropTypes.object.isRequired,
    
};

export default TodoList;
