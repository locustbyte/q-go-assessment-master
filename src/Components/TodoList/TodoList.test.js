import React from 'react';
import renderer from 'react-test-renderer';

import ReactDOM from 'react-dom';
import TodoList from './TodoList';

import uuid from 'uuid';
import moment from 'moment';

const todos = [{
    title: "Call the evangelist",
    completed: true,
    added: moment().toISOString(),
    category: 2,
    id: uuid.v4()
}];

const categories = [
    "Private",
    "Work",
    "Free time",
    "School",
    "Shopping"
];

function createNodeMock(element) {
    return '[ref object]';
}

describe('<TodoList />', () =>{

    it('should match its empty snapshot', () => {
        const tree = renderer.create(
            <TodoList />, {createNodeMock}
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
