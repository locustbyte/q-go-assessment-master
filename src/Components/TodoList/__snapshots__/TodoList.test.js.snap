exports[`<TodoList /> should match its empty snapshot 1`] = `
<div
  className="itemCreator">
  <form
    onSubmit={[Function]}>
    <input
      autoFocus={true}
      className="itemCreator-input"
      onChange={[Function]}
      placeholder="What do you need to do?"
      type="text"
      value="" />
    <input
      className="itemCreator-button"
      type="submit"
      value="Add Task" />
  </form>
  <ul
    className="itemsList-ul">
    <li
      className="completed-false">
      <input
        onClick={[Function]}
        type="checkbox" />
      <strong>
        TASK:
      </strong>
       
      Make sure items are completeable
       
      <span
        className="added">
        added 
        a month ago
      </span>
       
      <span
        onClick={[Function]}>
        <i
          aria-hidden="true"
          className="fa fa-minus-circle" />
         DELETE
      </span>
    </li>
    <li
      className="completed-false">
      <input
        onClick={[Function]}
        type="checkbox" />
      <strong>
        TASK:
      </strong>
       
      Add filters (Use HOC)
       
      <span
        className="added">
        added 
        a month ago
      </span>
       
      <span
        onClick={[Function]}>
        <i
          aria-hidden="true"
          className="fa fa-minus-circle" />
         DELETE
      </span>
    </li>
  </ul>
  <ul
    className="filters">
    <li>
      Showing: 
    </li>
    <li
      className="active"
      onClick={[Function]}>
      ALL
    </li>
    <li
      className=""
      onClick={[Function]}>
      COMPLETE
    </li>
    <li
      className=""
      onClick={[Function]}>
      INCOMPLETE
    </li>
  </ul>
  <div
    className="instructions">
    <h2>
      To run tests: npm test
    </h2>
    <hr />
    <h2>
      Production Build: yarn build (npm install yarn)
    </h2>
    <h4>
      (Produce minified scripts etc)
    </h4>
    <hr />
  </div>
</div>
`;
